api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,207 @@
+	ContentType        string                 `json:"contentType"`
+	Extensions         map[string]interface{} `json:"extensions"`
+	Data               interface{}            `json:"data"` // from docs: the payload is encoded into a media format which is specified by the contentType attribute (e.g. application/json)
+}
+
+type cloudEventIn struct {
+	CloudEvent
+
+	// Deadline string `json:"deadline"`
+	// Protocol CallRequestHTTP `json:"protocol"`
+}
+
+// cloudEventOut the expected response from the function container
+type cloudEventOut struct {
+	CloudEvent
+
+	// Protocol    *CallResponseHTTP `json:"protocol,omitempty"`
+}
+
+// CloudEventProtocol converts stdin/stdout streams from HTTP into JSON format.
+type CloudEventProtocol struct {
+	// These are the container input streams, not the input from the request or the output for the response
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *CloudEventProtocol) IsStreamable() bool {
+	return true
+}
+
+func (h *CloudEventProtocol) writeJSONToContainer(ci CallInfo) error {
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufPool.Put(buf)
+
+	_, err := io.Copy(buf, ci.Input())
+	if err != nil {
+		return err
+	}
+
+	var in cloudEventIn
+	// fmt.Printf("is cloudevent? %v\n", ci.IsCloudEvent())
+	if ci.IsCloudEvent() {
+		// then it's already in the right format, let's parse it, then modify
+		err = json.Unmarshal(buf.Bytes(), &in)
+		if err != nil {
+			return fmt.Errorf("Invalid CloudEvent input. %v", err)
+		}
+	} else {
+		in = cloudEventIn{
+			CloudEvent: CloudEvent{
+				ContentType:        ci.ContentType(),
+				EventID:            ci.CallID(),
+				EventType:          "http",
+				CloudEventsVersion: "0.1",
+				Source:             ci.RequestURL(),
+			},
+		}
+		if buf.Len() == 0 {
+			// nada
+			// todo: should we leave as null, pass in empty string, omitempty or some default for the content type, eg: {} for json?
+		} else if ci.ContentType() == "application/json" {
+			d := map[string]interface{}{}
+			err = json.NewDecoder(buf).Decode(&d)
+			if err != nil {
+				return fmt.Errorf("Invalid json body with contentType 'application/json'. %v", err)
+			}
+			in.Data = d
+		} else {
+			in.Data = buf.String()
+		}
+	}
+	// todo: deal with the dual ID's, one from outside, one from inside
+	if in.Extensions == nil {
+		in.Extensions = map[string]interface{}{}
+	}
+	// note: protocol stuff should be set on first ingestion of the event in fn2.0, the http router for example, not here
+	in.Extensions["protocol"] = CallRequestHTTP{
+		Type:       ci.ProtocolType(),
+		Method:     ci.Method(),
+		RequestURL: ci.RequestURL(),
+		Headers:    ci.Headers(),
+	}
+	in.Extensions["deadline"] = ci.Deadline().String()
+
+	fmt.Printf("SENDING IN: %+v\n", in)
+	return json.NewEncoder(h.in).Encode(in)
+}
+
+func (h *CloudEventProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) error {
+	ctx, span := trace.StartSpan(ctx, "dispatch_json")
```

Comment:
s/dispatch_json/dispatch_cloudevent/g (there's a few in this file)

Reply:
Fixed.